internal/consumer/dlq_test.go

```diff
@@ -0,0 +1,110 @@
+package consumer
+
+import (
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/stretchr/testify/suite"
+	"github.com/uber-go/kafka-client/internal/util"
+	"github.com/uber-go/tally"
+	"go.uber.org/zap"
+)
+
+type DLQTestSuite struct {
+	suite.Suite
+	saramaProducer *mockSaramaProducer
+	dlq            *bufferedDLQImpl
+	timeLimit      chan time.Time
+	sizeLimit      chan time.Time
+}
+
+func TestDLQTestSuite(t *testing.T) {
+	suite.Run(t, new(DLQTestSuite))
+}
+
+func (s *DLQTestSuite) SetupTest() {
+	s.timeLimit = make(chan time.Time)
+	s.sizeLimit = make(chan time.Time)
+	s.saramaProducer = newMockSaramaProducer()
+	logger := zap.NewNop()
+	s.dlq = &bufferedDLQImpl{
+		topic:            "topic",
+		producer:         s.saramaProducer,
+		timeLimit:        s.timeLimit,
+		sizeLimit:        s.sizeLimit,
+		stopC:            make(chan struct{}),
+		closedC:          make(chan struct{}),
+		messageBufferMap: newMessageBufferMap(),
+		errorMap:         newErrorMap(),
+		scope:            tally.NoopScope,
+		logger:           logger,
+		lifecycle:        util.NewRunLifecycle("dlqTest", logger),
+	}
+}
+
+func (s *DLQTestSuite) TestBatchProducer() {
+	m1 := &mockMessage{
+		topic:     "topic",
+		partition: 0,
+		offset:    0,
+	}
+	m2 := &mockMessage{
+		topic:     "topic",
+		partition: 0,
+		offset:    1,
+	}
+
+	// fail to produce m2
+	pe := &sarama.ProducerError{
+		Msg: &sarama.ProducerMessage{
+			Metadata: kafkaMessageKey(m2),
+		},
+		Err: errors.New("error"),
+	}
+	s.saramaProducer.errs = append(s.saramaProducer.errs, pe)
+	go s.dlq.eventLoop()
+
+	thread1 := make(chan error)
+	go func() {
+		err := s.dlq.Add(m1)
```

Comment:
also need a test case to verify that senders blocked on Add() are woken up when dlq is shutdown

Reply:
Added a test.